contracts/distribution/subdomains/ZNSSubdomainRegistrar.sol

```diff
@@ -0,0 +1,229 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.18;
+
+import { AZNSPayment } from "./abstractions/AZNSPayment.sol";
+import { AZNSPricing } from "./abstractions/AZNSPricing.sol";
+import { AZNSPricingWithFee } from "./abstractions/AZNSPricingWithFee.sol";
+import { AZNSRefundablePayment } from "./abstractions/AZNSRefundablePayment.sol";
+import { IZNSRegistry } from "../../registry/IZNSRegistry.sol";
+import { IZNSRegistrar } from "../IZNSRegistrar.sol";
+import { IZNSSubdomainRegistrar } from "./IZNSSubdomainRegistrar.sol";
+import { AAccessControlled } from "../../access/AAccessControlled.sol";
+import { ARegistryWired } from "../../abstractions/ARegistryWired.sol";
+
+
+contract ZNSSubdomainRegistrar is AAccessControlled, ARegistryWired, IZNSSubdomainRegistrar {
+    // TODO sub: change name of Registrar contract
+    IZNSRegistrar public rootRegistrar;
+
+    // TODO sub: make better name AND for the setter function !
+    mapping(bytes32 domainHash => DistributionConfig) public distrConfigs;
+
+    mapping(bytes32 domainHash =>
+        mapping(address registrant => bool allowed)
+    ) public distributionWhitelist;
+
+    modifier onlyOwnerOperatorOrRegistrar(bytes32 domainHash) {
+        require(
+            registry.isOwnerOrOperator(domainHash, msg.sender)
+            || accessController.isRegistrar(msg.sender),
+            "ZNSSubdomainRegistrar: Not authorized"
+        );
+        _;
+    }
+
+    // TODO sub: proxy ??
+    constructor(
+        address _accessController,
+        address _registry,
+        address _rootRegistrar
+    ) {
+        _setAccessController(_accessController);
+        setRegistry(_registry);
+        setRootRegistrar(_rootRegistrar);
+    }
+
+    function registerSubdomain(
+        bytes32 parentHash,
+        string calldata label,
+        address domainAddress,
+        DistributionConfig calldata distrConfig
+    ) external override returns (bytes32) {
+        // TODO sub: make the order of ops better
+        DistributionConfig memory parentConfig = distrConfigs[parentHash];
```

Comment:
could you please elaborate?
I don't think solc optimizes anything here. Usually every time you read from state var you pay an SSTORE, even if you are reading the same var, no?
we are reading this struct 6 times here. if I do not use memory var, I am paying for 6 SSTOREs, instead of 1 SSTORE and 5 MSTOREs. did I miss some kind of new optimization of solc?

Reply:
This has been a thing in Solidity since forever and is one of the main advices on how to reduce gas costs - using memory vars for multiple state reads. But I can test it, just to be sure.

Reply:
This is interesting. I must've missed this update. The memory var usage was always the way to go.
It still seems like the way to go, but I was surprised by the result of testing.
I did 2 ways of testing: estimateGas from ethers.js and running it on tenderly evm. The results are different and I'm not sure why thh diff is so big between those 2 methods.

Results:
memory var: 222002 gas - ethers (336,855 gas - tenderly)
direct state reads: 222427 gas - ethers (337,276 gas - tenderly)

We still see that using the var reduces the gas by around 400gas. Which is a very small amount. And this is strange.
Could you elaborate (if you know), how exactly these state reads are optimized now?

Reply:
For anyone else:
This has been discussed and researched.
"Each storage read (opcode sload) of the same slot costs 2,100 gas the first time during a transaction, and then 100 gas each other time during the same transaction. (After the EIP-2929 implemented in the Berlin hardfork. Before that, it was 800 for each read no matter how many times you performed the reads.)" (c)

What Damien also meant in above comments is that creating a memory var can reduce stack size (correct me if I'm wrong). Want to specify here that memory and stack are different things. We often see an error like "stack too deep, try reducing local variables". The limitation comes not from the stack size (which is 1024 elements), but from some opcodes being able to handle up to 16 elements (SWAP, DUP, etc.). This creates not a problem of stack being too populated, but the way local memory vars are structured and used within a *single function*. Meaning, a single function can only hold up to 16 variables.
While I am still not sure how much of the stack persists between functions, 16 vars in a total callstack is not a limit anymore, since stack (if not persisted) would be swapped to put newer vars on top. Since memory does not persist between function calls, the 16 var limit is not the problem anymore. It is the availability of a variable in the stack (being in the top 16 elements).

So all that is happening here is using memory to reduce state reads (SLOAD - 2100/100 gas) with memory reads (MLOAD - 3 gas). @durienb claims that in this case we do not use SLOAD 6 times, but only use once and then use it from the stack. I am not sure this is true, couldn't verify yet.
If we use approx calculation in gas diff, we get:
Only state reads: 2100 + 100*5 = 2600 gas (for 6 total reads)
State read with memory var: 2100 + 3*6 = 2118 gas (with one MSTORE)
26000 - 2118 = 482 gas - which is suspiciously close to 500 gas for 5 extra SLOADs, but this is not confirmed.

Conslusion:
Since we do not have a stack problem or too many vars in this function, we can keep the memory var and save about 500 gas here. The other option can be done, but deemed unnecessary for now.